provisioner/aws-pricing.js

```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+var Promise = require('promise');
+var lodash = require('lodash');
+var util = require('util');
+var assert = require('assert');
+var debug = require('debug')('aws-provisioner:aws-pricing');
+
+
+function fetchPricing(ec2) {
+  var that = this;
+  var startDate = new Date();
+  startDate.setMinutes(startDate.getMinutes() - 30);
+  var requestObj = {
+    StartTime: startDate,
+    Filters: [{
+      Name: 'product-description',
+      Values: ['Linux/UNIX'],
+    }],
+  }
+
+  var p = Promise.all([
+    ec2.describeSpotPriceHistory(requestObj),
+    ec2.describeAvailabilityZones({
+      Filters: [{
+        Name: 'state',
+        Values: ['available'],
+      }],
+    }),
+  ]);
+
+  p = p.then(function(res) {
+    var pricing = res[0];
+    var availableAvailZones = res[1]
+    var regions = Object.keys(pricing);
+    var fixed = {};
+    var zoneInfo = {};
+
+    // Get rid of the key we don't care about
+    regions.forEach(function(region) {
+      fixed[region] = pricing[region].SpotPriceHistory;
+      zoneInfo[region] = availableAvailZones[region].AvailabilityZones.map(function(x) {
+        return x.ZoneName;
+      });
+    });
+
+
+    return new AwsPricing(ec2.regions, fixed, zoneInfo);
+  })
+
+  return p;
+}
+
+module.exports = fetchPricing;
+
+function AwsPricing(regions, pricing, zoneInfo) {
+  assert(regions);
+  assert(pricing);
+  assert(zoneInfo);
+  this.__pricing = pricing;
+  this.regions = regions;
+  this.__zoneInfo = zoneInfo;
+}
+
+
+/**
+ * Get the pricing dictionary... This is a little ugly
+ * but it works... Ideas welcomed
+ */
+AwsPricing.prototype.get = function () {
+  return this.__pricing;
+};
+
+/**
+ * List the available availability zones
+ */
+AwsPricing.prototype.availableAvailabilityZones = function() {
+  var that = this;
+  var zones = [];
+  Object.keys(this.__zoneInfo).forEach(function(region) {
+    Array.prototype.push.apply(zones, that.availableAvailabilityZonesInRegion(region)); 
+  });
+  return zones;
+};
+
+
+/**
+ * List the available availability zones in one region
+ */
+AwsPricing.prototype.availableAvailabilityZonesInRegion = function(region) {
```

Comment:
I strongly suspect that `availabilityZonesInRegion` is explicit enough... Why would we ever want to list anything not available...